### ISSUE TITLE
Model trained without sample_weight and with sample_weight (where all sample weights = 1) gives different results

### ISSUE THREAD

**Issue by vlawhern** [skorch-dev/skorch#630]
Hello all,

I'm interested in using skorch to build classifiers where there are significant class imbalances. Per the [FAQ](https://skorch.readthedocs.io/en/stable/user/FAQ.html#i-want-to-use-sample-weight-how-can-i-do-this
), you can do this by passing in a dict with keys 'data' and 'sample_weight'. However, when I do this I'm getting very different results even when setting sample weights of all samples = 1 (so this should be no different than no weights at all).

See code below for example:

```python

import numpy as np
from sklearn.datasets import make_classification
from torch import nn
import torch.nn.functional as F

from skorch import NeuralNetClassifier, NeuralNet



################### Unweighted Version #######################################
X, y = make_classification(1000, 20, n_informative=10, random_state=0)
X = X.astype(np.float32)
y = y.astype(np.int64)

class MyModule(nn.Module):
    def __init__(self, num_units=10, nonlin=F.relu):
        super(MyModule, self).__init__()

        self.dense0 = nn.Linear(20, num_units)
        self.nonlin = nonlin
        self.dropout = nn.Dropout(0.5)
        self.dense1 = nn.Linear(num_units, 10)
        self.output = nn.Linear(10, 2)

    def forward(self, X):
        X = self.nonlin(self.dense0(X))
        X = self.dropout(X)
        X = F.relu(self.dense1(X))
        X = F.softmax(self.output(X), dim=-1)
        return X



net = NeuralNetClassifier(
    MyModule,
    max_epochs=10,
    lr=0.1,
    # Shuffle training data on each epoch
    iterator_train__shuffle=True,
)

net.fit(X, y)

  epoch    train_loss    valid_acc    valid_loss     dur
-------  ------------  -----------  ------------  ------
      1        0.6990       0.5550        0.6855  0.0370
      2        0.6866       0.5950        0.6781  0.0192
      3        0.6773       0.6350        0.6702  0.0149
      4        0.6707       0.6650        0.6613  0.0146
      5        0.6606       0.6700        0.6537  0.0148
      6        0.6508       0.6750        0.6454  0.0153
      7        0.6414       0.6800        0.6385  0.0147
      8        0.6368       0.6900        0.6334  0.0152
      9        0.6147       0.6800        0.6287  0.0146
     10        0.6168       0.6850        0.6208  0.0149


################## Weighted Version, with sample_weights = 1 #################

# put your X into a dict if not already a dict
X1 = {'data': X}
# add sample_weight to the X dict. All ones so should get similar result as above
X1['sample_weight'] = np.ones(shape=(1000,))

# construct new module to add 'data' and 'sample_weight' to forward pass
# https://skorch.readthedocs.io/en/stable/user/FAQ.html#i-want-to-use-sample-weight-how-can-i-do-this

class MyModule2(nn.Module):
    def __init__(self, num_units=10, nonlin=F.relu):
        super(MyModule2, self).__init__()

        self.dense0 = nn.Linear(20, num_units)
        self.nonlin = nonlin
        self.dropout = nn.Dropout(0.5)
        self.dense1 = nn.Linear(num_units, 10)
        self.output = nn.Linear(10, 2)

    def forward(self, data, sample_weight):
        X = self.nonlin(self.dense0(data))
        X = self.dropout(X)
        X = F.relu(self.dense1(X))
        X = F.softmax(self.output(X), dim=-1)
        return X

class MyNet(NeuralNet):
    def __init__(self, *args, criterion__reduce=False, **kwargs):
        # make sure to set reduce=False in your criterion, since we need the loss
        # for each sample so that it can be weighted
        super().__init__(*args, criterion__reduce=criterion__reduce, **kwargs)

    def get_loss(self, y_pred, y_true, X, *args, **kwargs):
         # override get_loss to use the sample_weight from X
         loss_unreduced = super().get_loss(y_pred, y_true, X, *args, **kwargs)
         sample_weight = X['sample_weight']
         loss_reduced = (sample_weight * loss_unreduced).mean()
         return loss_reduced

# construct new net. have to pass in criterion otherwise it throws an error
net1 = MyNet(
    MyModule2,
    max_epochs=10,
    lr=0.1,
    # Shuffle training data on each epoch
    iterator_train__shuffle=True,
    criterion = nn.NLLLoss)

net1.fit(X1, y)

  epoch    train_loss    valid_loss     dur
-------  ------------  ------------  ------
      1       -0.4932       -0.5179  0.0226
      2       -0.5005       -0.5195  0.0213
      3       -0.5030       -0.5224  0.0163
      4       -0.5097       -0.5248  0.0160
      5       -0.5121       -0.5281  0.0162
      6       -0.5192       -0.5324  0.0160
      7       -0.5224       -0.5369  0.0160
      8       -0.5361       -0.5414  0.0163
      9       -0.5429       -0.5461  0.0159
     10       -0.5456       -0.5515  0.0161

**Comment by BenjaminBossan** [comment]
Hey @vlawhern, the reason is that in the second case, you use `NeuralNet` instead of `NeuralNetClassifier`. When you replace it (at which point you don't need to specify the criterion anymore), you should get the same results.

Just as an explanation: `NeuralNetClassifier` applies a log on the softmax output of the module, but `NeuralNet` doesn't. That's why you get negative losses.

If this solves your problem, feel free to close the issue.

**Comment by vlawhern** [comment]
Yes this solves my problem (difference between ``NeuralNet`` and ``NeuralNetClassifier``). So I'll go ahead and close the issue. Thanks again!
